agent/consul/leader_connect.go

```diff
@@ -371,6 +371,20 @@ func (s *Server) initializeSecondaryCA(provider ca.Provider, roots structs.Index
 		needsNewIntermediate = true
 	}
 
+	if activeIntermediate != "" {
+		intermediateCert, err := connect.ParseCert(activeIntermediate)
+		if err != nil {
+			return fmt.Errorf("error parsing active intermediate cert: %v", err)
+		}
+
+		// Get a new intermediate from the primary DC if the existing
+		// intermediate is due for renew.
+		if moreThanHalfTimeHasPassed(time.Now(), intermediateCert.NotBefore,
+			intermediateCert.NotAfter) {
+			needsNewIntermediate = true
+		}
+	}
```

Comment:
This currently relies on our blocking query implementation which times out every 10 mins to implement a "periodic check". I'm not sure that's the best approach for a few reasons:

 1. When we upgrade this to streaming there will be no periodic timeouts
 2. We are using an implementation detail (that this method is called regularly even when there are no changes) in an unexpected/undocumented way. Even without streaming it would be easy to miss this assumption and for example add a line to the replication loop that skips calling this in the case of a timeout where the index is the same. I doubt that's plausibly going to happen in practice but is an example of why it feels wrong to implicitly assume regular iterations of this loop in the steady state.

Could we instead have a dedicated timer loop that somehow triggers the renewal? I realise it's more complicated having two different goroutines call this as they could do so concurrently but I don't think it's that hard to either trigger it via the same goroutine as now - just have it select on either watch update or timer update - or else jsut synchronize them with a mutex?

Reply:
Thats a good point. It required me redoing most of the work, but it is better now!